File: backend/app/schemas/subcontracting.py
"""Pydantic v2 schemas for the Subcontracting System."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict

from app.models.subcontracting import (
    SCAIAgentType, SCIssueStatus, SCOrderStatus,
    SCReceiptStatus, SCYieldStatus, ScrapReasonCode,
)


# ── SubcontractorLocation ─────────────────────────────────────────────────────

class SubcontractorLocationCreate(BaseModel):
    supplier_id:  uuid.UUID
    warehouse_id: uuid.UUID
    notes:        Optional[str] = None


class SubcontractorLocationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:           uuid.UUID
    supplier_id:  uuid.UUID
    warehouse_id: uuid.UUID
    is_active:    bool
    notes:        Optional[str]
    created_at:   Optional[datetime]
    supplier_name:  Optional[str] = None
    warehouse_name: Optional[str] = None
    warehouse_code: Optional[str] = None


# ── Order ─────────────────────────────────────────────────────────────────────

class SCOrderLineCreate(BaseModel):
    line_no:            int
    product_id:         Optional[uuid.UUID] = None
    material_id:        Optional[uuid.UUID] = None
    description:        Optional[str] = None
    quantity_ordered:   Decimal
    uom:                str = "KG"
    bom_id:             Optional[uuid.UUID] = None
    service_unit_cost:  Optional[Decimal] = None
    estimated_yield_pct: Optional[Decimal] = Decimal("100")
    notes:              Optional[str] = None


class SCOrderCreate(BaseModel):
    supplier_id:              uuid.UUID
    order_date:               date
    expected_completion_date: Optional[date] = None
    warehouse_id:             Optional[uuid.UUID] = None
    linked_po_id:             Optional[uuid.UUID] = None
    currency:                 str = "KES"
    remarks:                  Optional[str] = None
    lines:                    List[SCOrderLineCreate] = []


class SCOrderLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:                  uuid.UUID
    order_id:            uuid.UUID
    line_no:             int
    product_id:          Optional[uuid.UUID]
    material_id:         Optional[uuid.UUID]
    description:         Optional[str]
    quantity_ordered:    Decimal
    quantity_received:   Decimal
    uom:                 str
    bom_id:              Optional[uuid.UUID]
    service_unit_cost:   Optional[Decimal]
    estimated_yield_pct: Optional[Decimal]
    notes:               Optional[str]
    product_name:  Optional[str] = None
    material_name: Optional[str] = None


class SCOrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:                       uuid.UUID
    order_no:                 str
    supplier_id:              uuid.UUID
    order_date:               date
    expected_completion_date: Optional[date]
    actual_completion_date:   Optional[date]
    status:                   SCOrderStatus
    linked_po_id:             Optional[uuid.UUID]
    warehouse_id:             Optional[uuid.UUID]
    subcontractor_location_id: Optional[uuid.UUID]
    total_material_cost:      Optional[Decimal]
    total_service_cost:       Optional[Decimal]
    total_wastage_cost:       Optional[Decimal]
    currency:                 str
    approved_by_id:           Optional[uuid.UUID]
    approved_at:              Optional[datetime]
    created_by_id:            Optional[uuid.UUID]
    remarks:                  Optional[str]
    created_at:               Optional[datetime]
    supplier_name:  Optional[str] = None
    warehouse_name: Optional[str] = None
    lines:          List[SCOrderLineOut] = []


# ── Material Issue ────────────────────────────────────────────────────────────

class SCIssueLineCreate(BaseModel):
    line_no:          int
    material_id:      uuid.UUID
    lot_id:           Optional[uuid.UUID] = None
    quantity_issued:  Decimal
    uom:              str = "KG"
    unit_cost:        Optional[Decimal] = None
    notes:            Optional[str] = None


class SCIssueCreate(BaseModel):
    order_id:   uuid.UUID
    issue_date: date
    notes:      Optional[str] = None
    lines:      List[SCIssueLineCreate] = []


class SCIssueLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:                uuid.UUID
    issue_id:          uuid.UUID
    line_no:           int
    material_id:       uuid.UUID
    lot_id:            Optional[uuid.UUID]
    quantity_issued:   Decimal
    quantity_returned: Decimal
    quantity_consumed: Decimal
    quantity_scrapped: Decimal
    uom:               str
    unit_cost:         Optional[Decimal]
    stock_movement_id: Optional[uuid.UUID]
    notes:             Optional[str]
    material_name: Optional[str] = None
    material_code: Optional[str] = None
    lot_number:    Optional[str] = None


class SCIssueOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:          uuid.UUID
    issue_no:    str
    order_id:    uuid.UUID
    issue_date:  date
    status:      SCIssueStatus
    issued_by_id: Optional[uuid.UUID]
    notes:       Optional[str]
    created_at:  Optional[datetime]
    lines:       List[SCIssueLineOut] = []


# ── Receipt ───────────────────────────────────────────────────────────────────

class SCReceiptLineCreate(BaseModel):
    line_no:           int
    order_line_id:     Optional[uuid.UUID] = None
    product_id:        Optional[uuid.UUID] = None
    material_id:       Optional[uuid.UUID] = None
    quantity_received: Decimal
    quantity_accepted: Optional[Decimal] = None
    quantity_rejected: Decimal = Decimal("0")
    uom:               str = "KG"
    lot_number:        Optional[str] = None
    expiry_date:       Optional[date] = None
    unit_service_cost: Optional[Decimal] = None
    notes:             Optional[str] = None


class SCReceiptCreate(BaseModel):
    order_id:          uuid.UUID
    receipt_date:      date
    qc_inspection_id:  Optional[uuid.UUID] = None
    notes:             Optional[str] = None
    lines:             List[SCReceiptLineCreate] = []


class SCReceiptLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:                uuid.UUID
    receipt_id:        uuid.UUID
    order_line_id:     Optional[uuid.UUID]
    line_no:           int
    product_id:        Optional[uuid.UUID]
    material_id:       Optional[uuid.UUID]
    quantity_received: Decimal
    quantity_accepted: Decimal
    quantity_rejected: Decimal
    uom:               str
    lot_number:        Optional[str]
    expiry_date:       Optional[date]
    unit_service_cost: Optional[Decimal]
    stock_movement_id: Optional[uuid.UUID]
    notes:             Optional[str]
    product_name:  Optional[str] = None
    material_name: Optional[str] = None


class SCReceiptOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:               uuid.UUID
    receipt_no:       str
    order_id:         uuid.UUID
    receipt_date:     date
    status:           SCReceiptStatus
    received_by_id:   Optional[uuid.UUID]
    qc_inspection_id: Optional[uuid.UUID]
    notes:            Optional[str]
    created_at:       Optional[datetime]
    lines:            List[SCReceiptLineOut] = []


# ── Yield Record ──────────────────────────────────────────────────────────────

class SCYieldOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:                      uuid.UUID
    order_id:                uuid.UUID
    order_line_id:           uuid.UUID
    total_material_issued:   Decimal
    expected_material_input: Optional[Decimal]
    quantity_ordered:        Decimal
    quantity_received:       Decimal
    expected_yield_pct:      Optional[Decimal]
    actual_yield_pct:        Optional[Decimal]
    yield_variance_pct:      Optional[Decimal]
    yield_status:            SCYieldStatus
    total_scrapped:          Decimal
    scrap_reason:            Optional[ScrapReasonCode]
    scrap_cost:              Optional[Decimal]
    material_variance_qty:   Optional[Decimal]
    material_variance_cost:  Optional[Decimal]
    is_abnormal:             bool
    notes:                   Optional[str]


# ── Performance ───────────────────────────────────────────────────────────────

class SCPerformanceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:                  uuid.UUID
    order_id:            uuid.UUID
    supplier_id:         uuid.UUID
    planned_completion:  Optional[date]
    actual_completion:   Optional[date]
    delay_days:          Optional[int]
    on_time:             Optional[bool]
    total_qty_ordered:   Decimal
    total_qty_received:  Decimal
    total_qty_rejected:  Decimal
    rejection_rate_pct:  Optional[Decimal]
    avg_yield_pct:       Optional[Decimal]
    budgeted_cost:       Optional[Decimal]
    actual_cost:         Optional[Decimal]
    cost_variance_pct:   Optional[Decimal]
    performance_score:   Optional[Decimal]
    notes:               Optional[str]
    supplier_name: Optional[str] = None
    order_no:      Optional[str] = None


# ── AI Rec ────────────────────────────────────────────────────────────────────

class SCAIRecOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:               uuid.UUID
    order_id:         Optional[uuid.UUID]
    supplier_id:      Optional[uuid.UUID]
    agent_type:       SCAIAgentType
    title:            str
    recommendation:   str
    rationale:        Optional[str]
    risk_level:       Optional[str]
    potential_saving: Optional[Decimal]
    priority:         int
    is_actioned:      bool
    action_notes:     Optional[str]
    created_at:       Optional[datetime]
    supplier_name: Optional[str] = None
    order_no:      Optional[str] = None


# ── Subcontractor Stock View ──────────────────────────────────────────────────

class SubcontractorStockRow(BaseModel):
    supplier_id:    uuid.UUID
    supplier_name:  str
    material_id:    uuid.UUID
    material_code:  str
    material_name:  str
    uom:            Optional[str]
    qty_issued:     Decimal
    qty_returned:   Decimal
    qty_consumed:   Decimal
    qty_scrapped:   Decimal
    qty_balance:    Decimal   # issued - returned - consumed - scrapped
    unit_cost:      Optional[Decimal]
    total_value:    Optional[Decimal]


# ── Dashboard ─────────────────────────────────────────────────────────────────

class SCDashboard(BaseModel):
    total_orders:        int
    draft_orders:        int
    active_orders:       int
    completed_orders:    int
    overdue_orders:      int
    total_material_issued_value: Decimal
    avg_yield_pct:       Optional[Decimal]
    total_scrap_cost:    Decimal
    ai_recs_count:       int
    supplier_count:      int
    locations_count:     int

File: frontend/src/lib/subcontracting.ts
import { apiClient } from "@/lib/api";

// ── Enums ─────────────────────────────────────────────────────────────────────
export type SCOrderStatus = "DRAFT"|"APPROVED"|"ISSUED"|"IN_PROGRESS"|"PARTIALLY_RECEIVED"|"COMPLETED"|"CLOSED"|"CANCELLED";
export type SCIssueStatus = "DRAFT"|"ISSUED"|"PARTIALLY_RETURNED"|"RETURNED";
export type SCReceiptStatus = "DRAFT"|"POSTED"|"QC_HOLD"|"REJECTED";
export type SCYieldStatus = "NORMAL"|"LOW_YIELD"|"HIGH_SCRAP"|"ABNORMAL";
export type SCAIAgentType = "PERFORMANCE_ANALYZER"|"COST_OPTIMIZER"|"RISK_DETECTOR";

// ── Types ─────────────────────────────────────────────────────────────────────
export interface SCDashboard {
  total_orders: number; draft_orders: number; active_orders: number;
  completed_orders: number; overdue_orders: number;
  total_material_issued_value: number; avg_yield_pct: number|null;
  total_scrap_cost: number; ai_recs_count: number;
  supplier_count: number; locations_count: number;
}
export interface SubcontractorLocationOut {
  id: string; supplier_id: string; warehouse_id: string;
  is_active: boolean; notes: string|null; created_at: string|null;
  supplier_name: string|null; warehouse_name: string|null; warehouse_code: string|null;
}
export interface SCOrderLineOut {
  id: string; order_id: string; line_no: number;
  product_id: string|null; material_id: string|null;
  description: string|null; quantity_ordered: number; quantity_received: number;
  uom: string; bom_id: string|null; service_unit_cost: number|null;
  estimated_yield_pct: number|null; notes: string|null;
  product_name: string|null; material_name: string|null;
}
export interface SCOrderOut {
  id: string; order_no: string; supplier_id: string; order_date: string;
  expected_completion_date: string|null; actual_completion_date: string|null;
  status: SCOrderStatus; linked_po_id: string|null; warehouse_id: string|null;
  subcontractor_location_id: string|null;
  total_material_cost: number|null; total_service_cost: number|null;
  total_wastage_cost: number|null; currency: string;
  approved_by_id: string|null; approved_at: string|null;
  created_by_id: string|null; remarks: string|null; created_at: string|null;
  supplier_name: string|null; warehouse_name: string|null;
  lines: SCOrderLineOut[];
}
export interface SCIssueLineOut {
  id: string; issue_id: string; line_no: number; material_id: string;
  lot_id: string|null; quantity_issued: number; quantity_returned: number;
  quantity_consumed: number; quantity_scrapped: number; uom: string;
  unit_cost: number|null; stock_movement_id: string|null; notes: string|null;
  material_name: string|null; material_code: string|null; lot_number: string|null;
}
export interface SCIssueOut {
  id: string; issue_no: string; order_id: string; issue_date: string;
  status: SCIssueStatus; issued_by_id: string|null; notes: string|null;
  created_at: string|null; lines: SCIssueLineOut[];
}
export interface SCReceiptLineOut {
  id: string; receipt_id: string; order_line_id: string|null; line_no: number;
  product_id: string|null; material_id: string|null;
  quantity_received: number; quantity_accepted: number; quantity_rejected: number;
  uom: string; lot_number: string|null; expiry_date: string|null;
  unit_service_cost: number|null; stock_movement_id: string|null; notes: string|null;
  product_name: string|null; material_name: string|null;
}
export interface SCReceiptOut {
  id: string; receipt_no: string; order_id: string; receipt_date: string;
  status: SCReceiptStatus; received_by_id: string|null;
  qc_inspection_id: string|null; notes: string|null; created_at: string|null;
  lines: SCReceiptLineOut[];
}
export interface SCYieldOut {
  id: string; order_id: string; order_line_id: string;
  total_material_issued: number; expected_material_input: number|null;
  quantity_ordered: number; quantity_received: number;
  expected_yield_pct: number|null; actual_yield_pct: number|null;
  yield_variance_pct: number|null; yield_status: SCYieldStatus;
  total_scrapped: number; scrap_reason: string|null; scrap_cost: number|null;
  material_variance_qty: number|null; material_variance_cost: number|null;
  is_abnormal: boolean; notes: string|null;
}
export interface SCPerformanceOut {
  id: string; order_id: string; supplier_id: string;
  planned_completion: string|null; actual_completion: string|null;
  delay_days: number|null; on_time: boolean|null;
  total_qty_ordered: number; total_qty_received: number; total_qty_rejected: number;
  rejection_rate_pct: number|null; avg_yield_pct: number|null;
  budgeted_cost: number|null; actual_cost: number|null;
  cost_variance_pct: number|null; performance_score: number|null;
  notes: string|null; supplier_name: string|null; order_no: string|null;
}
export interface SCAIRecOut {
  id: string; order_id: string|null; supplier_id: string|null;
  agent_type: SCAIAgentType; title: string; recommendation: string;
  rationale: string|null; risk_level: string|null; potential_saving: number|null;
  priority: number; is_actioned: boolean; action_notes: string|null;
  created_at: string|null; supplier_name: string|null; order_no: string|null;
}
export interface SubcontractorStockRow {
  supplier_id: string; supplier_name: string;
  material_id: string; material_code: string; material_name: string; uom: string|null;
  qty_issued: number; qty_returned: number; qty_consumed: number;
  qty_scrapped: number; qty_balance: number;
  unit_cost: number|null; total_value: number|null;
}

// ── Color helpers ─────────────────────────────────────────────────────────────
export const orderStatusBadge = (s: SCOrderStatus) => ({
  DRAFT:"bg-gray-100 text-gray-600", APPROVED:"bg-blue-100 text-blue-700",
  ISSUED:"bg-indigo-100 text-indigo-700", IN_PROGRESS:"bg-yellow-100 text-yellow-700",
  PARTIALLY_RECEIVED:"bg-orange-100 text-orange-700", COMPLETED:"bg-green-100 text-green-700",
  CLOSED:"bg-purple-100 text-purple-700", CANCELLED:"bg-red-100 text-red-700",
}[s] ?? "bg-gray-100 text-gray-600");

export const yieldBadge = (s: SCYieldStatus) => ({
  NORMAL:"bg-green-100 text-green-700", LOW_YIELD:"bg-orange-100 text-orange-700",
  HIGH_SCRAP:"bg-red-100 text-red-700", ABNORMAL:"bg-red-200 text-red-800",
}[s] ?? "bg-gray-100 text-gray-600");

export const riskBadge = (r: string|null) => ({
  LOW:"bg-green-100 text-green-700", MEDIUM:"bg-yellow-100 text-yellow-700",
  HIGH:"bg-orange-100 text-orange-700", CRITICAL:"bg-red-100 text-red-700",
}[r ?? ""] ?? "bg-gray-100 text-gray-600");

export const agentLabel = (a: SCAIAgentType) => ({
  PERFORMANCE_ANALYZER:"Performance Analyzer",
  COST_OPTIMIZER:"Cost Optimizer",
  RISK_DETECTOR:"Risk Detector",
}[a] ?? a);

export const fmt = (n: number|null|undefined, dp = 2) =>
  n == null ? "—" : Number(n).toLocaleString("en-KE", {minimumFractionDigits: dp, maximumFractionDigits: dp});

// ── API client ────────────────────────────────────────────────────────────────
const B = "/api/v1/subcontracting";

export const scApi = {
  getDashboard: () => apiClient.get<SCDashboard>(`${B}/dashboard`).then((r) => r.data),

  listLocations: () => apiClient.get<SubcontractorLocationOut[]>(`${B}/locations`).then((r) => r.data),
  createLocation: (p: {supplier_id: string; warehouse_id: string; notes?: string}) =>
    apiClient.post<SubcontractorLocationOut>(`${B}/locations`, p).then((r) => r.data),

  listOrders: (params?: {status?: string; supplier_id?: string; skip?: number; limit?: number}) => {
    const qs = new URLSearchParams();
    if (params?.status)      qs.set("status", params.status);
    if (params?.supplier_id) qs.set("supplier_id", params.supplier_id);
    if (params?.skip)        qs.set("skip", String(params.skip));
    if (params?.limit)       qs.set("limit", String(params.limit));
    return apiClient.get<SCOrderOut[]>(`${B}/orders?${qs}`).then((r) => r.data);
  },
  createOrder: (p: Partial<SCOrderOut> & {lines?: Partial<SCOrderLineOut>[]}) =>
    apiClient.post<SCOrderOut>(`${B}/orders`, p).then((r) => r.data),
  getOrder: (id: string) => apiClient.get<SCOrderOut>(`${B}/orders/${id}`).then((r) => r.data),
  approveOrder: (id: string) => apiClient.post<SCOrderOut>(`${B}/orders/${id}/approve`, {}).then((r) => r.data),
  completeOrder: (id: string) => apiClient.post<SCOrderOut>(`${B}/orders/${id}/complete`, {}).then((r) => r.data),

  issueMaterial: (orderId: string, p: Partial<SCIssueOut> & {lines?: Partial<SCIssueLineOut>[]}) =>
    apiClient.post<SCIssueOut>(`${B}/orders/${orderId}/issue-material`, p).then((r) => r.data),
  listIssues: (orderId: string) => apiClient.get<SCIssueOut[]>(`${B}/orders/${orderId}/issues`).then((r) => r.data),

  receiveGoods: (orderId: string, p: Partial<SCReceiptOut> & {lines?: Partial<SCReceiptLineOut>[]}) =>
    apiClient.post<SCReceiptOut>(`${B}/orders/${orderId}/receive`, p).then((r) => r.data),
  listReceipts: (orderId: string) => apiClient.get<SCReceiptOut[]>(`${B}/orders/${orderId}/receipts`).then((r) => r.data),

  getYield: (orderId: string) => apiClient.get<SCYieldOut[]>(`${B}/orders/${orderId}/yield`).then((r) => r.data),
  yieldReport: (abnormalOnly = false) =>
    apiClient.get<SCYieldOut[]>(`${B}/yield/report?abnormal_only=${abnormalOnly}`).then((r) => r.data),

  getStock: (supplierId?: string) => {
    const qs = supplierId ? `?supplier_id=${supplierId}` : "";
    return apiClient.get<SubcontractorStockRow[]>(`${B}/stock${qs}`).then((r) => r.data);
  },

  getPerformance: (supplierId?: string) => {
    const qs = supplierId ? `?supplier_id=${supplierId}` : "";
    return apiClient.get<SCPerformanceOut[]>(`${B}/performance${qs}`).then((r) => r.data);
  },

  runAI: (orderId?: string) => orderId
    ? apiClient.post<{generated: number; by_agent: Record<string, number>}>(`${B}/orders/${orderId}/ai/run`, {}).then((r) => r.data)
    : apiClient.post<{generated: number; by_agent: Record<string, number>}>(`${B}/ai/run`, {}).then((r) => r.data),
  listAIRecs: (params?: {order_id?: string; supplier_id?: string}) => {
    const qs = new URLSearchParams();
    if (params?.order_id)    qs.set("order_id", params.order_id);
    if (params?.supplier_id) qs.set("supplier_id", params.supplier_id);
    return apiClient.get<SCAIRecOut[]>(`${B}/ai/recommendations?${qs}`).then((r) => r.data);
  },
  actionAIRec: (id: string, notes?: string) =>
    apiClient.post<SCAIRecOut>(`${B}/ai/recommendations/${id}/action?notes=${encodeURIComponent(notes ?? "")}`, {}).then((r) => r.data),
};
